test/es6/classes.js

```diff
@@ -1140,6 +1133,55 @@ var tests = [
             assert.throws(function() { (new (class A { set x(_) { A = 0; } })).x = 15; }, TypeError, "Assignment to class identifier in setter");
         }
     },
+    {
+        name: "`class x extends y` where `y` is an expression containing identifier `x` should be a ReferenceError",
+        body: function() {
+            var errorText = "Use before declaration";
+
+            function assert_referr(code) {
+                assert.throws(function () { eval(code) }, ReferenceError, `\n    ${code}`, errorText);
+            }
+
+            function assert_referrors(code) {
+                assert_referr(code);
+                assert_referr(`var x = ${code}`);
+                assert_referr(`let x = ${code}`);
+            }
+
+            function id(x) { return x; };
+            function fun(x) { return function() { return x }; }
+
+            // In the Test262 test case, the RHS is a class expression.
+            // See: test262/test/language/statements/class/name-binding/in-extends-expression-assigned.js
+            assert_referrors("(class x extends x {})");
+
+            // Using expressions containing the `x` identifier for the extends clause
+            assert_referrors("class x extends x {}");
+            assert_referrors("class x extends (x) {}");
+            assert_referrors("class x extends id(x) {}");
+            assert_referrors("class x extends fun(x) {}");
+
+            // Assigning the result to a different identifier
+            assert_referr("var y = class x extends x {}");
+            assert_referr("let y = class x extends x {}");
+
+            // Define `y` after the class to use default initialization (var y) or temporary deadzone (let y).
+            assert.throws(`
+                class x extends y {}; // y == undefined
+                var y = function() {};
+            `, TypeError);

```

Comment:
@ianwjhalliday Fixed. It's supposed to be a TypeError but you're right that it was throwing TypeError for the wrong reason in this case.